### DISCUSSION TITLE
How to distinguish `(a?.().b!).c` and `(a?.().b)!.c`

### DISCUSSION THREAD

**Discussion by fisker** [babel/babel#15077] [General] [answered]
Them seems the same except `extra.parenthesized`, I ask this because `@typescript-eslint/typescript-estree` parse them differently.

`ChainExpression > TSNonNullExpression` vs  `TSNonNullExpression > ChainExpression `

I'm going to print `ChainExpression` in Prettier, but the AST compare won't pass if we follow the way we used print (`babel-ts`) https://github.com/prettier/prettier/pull/13735

Seems no way to distinguish them without checking `extra.parenthesized`?

**Comment by nicolo-ribaudo** [comment]
Have those two expressions a different meaning?

**Reply by fisker** [reply]
I don't know TS, as mentioned, only because `@typescript-eslint/typescript-estree` parse them differently. so I asked.

**Comment by nicolo-ribaudo** [comment, marked as answer]
These two expressions have the same behavior: https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEA3W8oC54DsCuELwB94AKYgSngF4A+eAb3gCMMc9D74wXsBbRkGPAC+wsgG4AUBIggALvAAeVElAD8AOnLrGAQjLqwkmfICey4ms37GZHQclA

In both cases `!` is type-asserting that `a?.().b` is not nullish. `!` is _not_ part of the optional chain (thus, asserting that `.b` is non-nullish only if `a` is non-nullish), otherwise the first example in the link above would report that `a?.().b` is possibly nullish (if `a` is nullish).

The parentheses here don't have any effect other than terminating the optional chain (thus making `.c` non-optional), thus there is no need to distinguish these two expressions from at an AST level. For comparison, it would be like distinguishing `() => ({}.a)` from `() => ({}).a`: the parentheses are necessary, but where they are doesn't matter.

If you _need_ to distinguish them you could use the `createParenthesizedExpressions` parser option, but Prettier could normalize them exactly as it normalizes `() => ({}.a)` and `() => ({}).a`.

**Reply by fisker** [reply]
Thank you!
